packages/backend/src/open_payments/payment/outgoing/service.ts

```diff
@@ -159,6 +185,326 @@ async function createOutgoingPayment(
+    return false
+  }
+  let {
+    estTotalAvailableSendAmount,
+    estTotalAvailableReceiveAmount
+  } = estimateTotalAvailableAmounts<PaymentAccess>(
+    paymentAccess,
+    payment,
+    deps.slippage
+  )
+
+  if (
+    estTotalAvailableSendAmount < payment.sendAmount.value ||
+    estTotalAvailableReceiveAmount < payment.receiveAmount.value
+  ) {
+    // Payment amount single-handedly exceeds sendAmount limit(s)
+    return false
+  }
+
+  // TODO: lock to prevent multiple grant payments from being created concurrently?
+  const grantPayments = await OutgoingPayment.query(deps.knex)
+    .where({
+      grant: grant.grant
+    })
+    .andWhereNot({
+      id: payment.id
+    })
+    .withGraphFetched('[quote.asset]')
+
+  if (grantPayments.length === 0) {
+    return true
+  }
+
+  const competingPayments: OutgoingPayment[] = []
+  for (const grantPayment of grantPayments) {
+    if (grantPayment.state === OutgoingPaymentState.Failed) {
+      const totalSent = await deps.accountingService.getTotalSent(
+        grantPayment.id
+      )
+      assert.ok(totalSent !== undefined)
+      if (totalSent === BigInt(0)) {
+        continue
+      }
+      grantPayment.sentAmount = { ...grantPayment.sentAmount, value: totalSent }
+    } else {
+      grantPayment.sentAmount = grantPayment.sendAmount
+    }
+    if (
+      paymentAccess.find((access) =>
+        validatePaymentAccess({
+          access,
+          payment: grantPayment,
+          identifier: `${deps.publicHost}/${grantPayment.accountId}`
+        })
+      ) &&
+      !unlimitedAccess.find((access) =>
+        validateAccess({
+          access,
+          payment: grantPayment,
+          identifier: `${deps.publicHost}/${grantPayment.accountId}`
+        })
+      )
+    ) {
+      competingPayments.push(grantPayment)
+    }
+  }
+  if (!competingPayments) {
+    // The payment may use the entire amount limit(s)
+    return true
+  }
+
+  // Do paymentAccess limits support payment and competing existing payments?
+  for (const grantPayment of competingPayments) {
+    let sentAmount = grantPayment.sentAmount.value
+    let receivedAmount: bigint
+    // Estimate delivered amount of failed payment
+    if (grantPayment.state === OutgoingPaymentState.Failed) {
+      receivedAmount =
+        (grantPayment.receiveAmount.value * grantPayment.sentAmount.value) /
+        grantPayment.sendAmount.value
+    } else {
+      receivedAmount = grantPayment.receiveAmount.value
+    }
+
+    // Check if competing payment was created with (for current payment) unrelated access
+    const competingUnrelatedAccess = unrelatedAccess.filter((access) =>
+      validateAccess({
+        access,
+        payment: grantPayment,
+        identifier: `${deps.publicHost}/${grantPayment.accountId}`
+      })
+    )
+    if (competingUnrelatedAccess.length > 0) {
+      for (const access of unrelatedAccess) {
+        if (competingUnrelatedAccess.indexOf(access) > -1) {
```

Comment:
Instead of populating and then searching `competingUnrelatedAccess`, could `validateAccess` be called at the start of this `for` loop?

Reply:
But then I always loop, even if there is no competing unrelated access. But that may still be faster than doing the filter operation. it gets rid of `competingUnrelatedAccess` so that's a plus.